src/vcpkg/base/downloads.cpp

```diff
@@ -883,6 +884,110 @@ namespace vcpkg
+        static const char* block_ids[10] = {
+            "MDBACg==",
+            "MDBBCg==",
+            "MDBCCg==",
+            "MDBDCg==",
+            "MDBECg==",
+            "MDBFCg==",
+            "MDBGCg==",
+            "MDBHCg==",
+            "MDBICg==",
+            "MDBJCg==",
+        };
+
+        if (file_size > block_size * std::extent<decltype(block_ids)>::value) return false;
+
+        auto http_cmd = Command{"curl"}.string_arg("--fail-early");
+
+        auto block_url = raw_url.to_string();
+        block_url += raw_url.contains('?') ? '&' : '?';
+        block_url += "&comp=block&blockid=";
+
+        auto block_file = fmt::format("{}.chunk", file);
+        auto block_id = std::begin(block_ids);
+        auto block_list = std::string("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<BlockList>\n");
+        for (size_t range_from = 0; range_from < file_size; range_from += block_size, ++block_id)
+        {
+            auto range_to = std::min(range_from + block_size, file_size);
+
+            // Extract block data locally
+            http_cmd.string_arg("-o").string_arg(block_file);
+            http_cmd.string_arg("--range").string_arg(fmt::format("{}-{}", range_from, range_to - 1));
+            http_cmd.string_arg(fmt::format("file://{}", file));
+            http_cmd.string_arg("--next");
+
+            // Upload block
+            add_curl_headers(http_cmd, headers);
+            http_cmd.string_arg("-X").string_arg("PUT");
+            http_cmd.string_arg(block_url + *block_id);
+            http_cmd.string_arg("-T").string_arg(block_file);
+            http_cmd.string_arg("--next");
+            http_cmd.string_arg("-w").string_arg("\\n" + guid_marker.to_string() + "%{http_code}");
+
+            block_list += "  <Latest>";
+            block_list += *block_id;
+            block_list += "</Latest>\n";
+        }
+        block_list += "</BlockList>\n";
+
+        std::error_code ec;
+        auto block_list_filepath = file + ".blocklist";
+        fs.write_contents(block_list_filepath, block_list, ec);
+        if (ec) return false;
+
+        auto block_list_url = raw_url.to_string();
+        block_list_url += raw_url.contains('?') ? '&' : '?';
+        block_list_url += "comp=blocklist";
+        add_curl_headers(http_cmd, {}); // do not send x-ms-blob-type!
+        http_cmd.string_arg("-H").string_arg("x-ms-version: 2020-04-08");
+        http_cmd.string_arg("-X").string_arg("PUT");
+        http_cmd.string_arg(block_list_url);
+        http_cmd.string_arg("-T").string_arg(block_list_filepath);
+        http_cmd.string_arg("-w").string_arg("\\n" + guid_marker.to_string() + "%{http_code}");
+
+        int code = 0;
+        auto res = cmd_execute_and_stream_lines(context, http_cmd, [&code](StringView line) {
+            if (line.starts_with(guid_marker))
+            {
+                code = std::strtol(line.data() + guid_marker.size(), nullptr, 10);
+            }
+        });
+
```

Comment:
Maybe block file and block list file should be removed at this point.

Reply:
I wounder that one can use the same output and input file for all request. I guess that only works when curl handles one request after each other which is the default?

Reply:
Before implementation, I read the man page for this question. AFAICT parallel mode is not enabled without extra options.

Reply:
It is doing `2n+1` operations in a single curl invocation: For each block, there is an operation to cut a block to a file and a second one to upload that file. In the end, there is the operation to upload the block list.

In parallel mode, it still couldn't upload a block before it is available as a file, it couldn't reuse the same filepath for the temporary block storage, and the blocklist would still have to wait for all blocks to be uploaded.

`azcopy` might be easier than `curl` if performance is the goal.